cfn-resources/stream-connection/cmd/resource/resource.go

```diff
@@ -0,0 +1,226 @@
+	util.SetupLogger("mongodb-atlas-stream-connection")
+
+	util.SetDefaultProfileIfNotDefined(&currentModel.Profile)
+
+	var requiredFields []string
+
+	switch cfnFunc {
+	case constants.CREATE:
+		requiredFields = CreateRequiredFields
+	case constants.READ:
+		requiredFields = ReadRequiredFields
+	case constants.UPDATE:
+		requiredFields = UpdateRequiredFields
+	case constants.DELETE:
+		requiredFields = DeleteRequiredFields
+	case constants.LIST:
+		requiredFields = ListRequiredFields
+	}
+
+	if errEvent := validator.ValidateModel(requiredFields, currentModel); errEvent != nil {
+		return nil, errEvent
+	}
+
+	client, peErr := util.NewAtlasV2OnlyClientLatest(&req, currentModel.Profile, true)
+	if peErr != nil {
+		return nil, peErr
+	}
+	return client.AtlasSDK, nil
+}
+
+func Create(req handler.Request, prevModel *Model, currentModel *Model) (handler.ProgressEvent, error) {
+	conn, peErr := setup(constants.CREATE, req, currentModel)
+	if peErr != nil {
+		return *peErr, nil
+	}
+
+	ctx := context.Background()
+
+	projectID := currentModel.ProjectId
+	instanceName := currentModel.InstanceName
+	streamConnectionReq := NewStreamConnectionReq(ctx, currentModel)
+
+	streamConnResp, apiResp, err := conn.StreamsApi.CreateStreamConnection(ctx, *projectID, *instanceName, streamConnectionReq).Execute()
+	if err != nil {
+		return handleError(apiResp, constants.CREATE, err)
+	}
+
+	resourceModel := GetStreamConnectionModel(streamConnResp, currentModel)
+
+	return handler.ProgressEvent{
+		OperationStatus: handler.Success,
+		Message:         "Create Completed",
+		ResourceModel:   resourceModel,
+	}, nil
+}
+
+func Read(req handler.Request, prevModel *Model, currentModel *Model) (handler.ProgressEvent, error) {
+	conn, peErr := setup(constants.READ, req, currentModel)
+	if peErr != nil {
+		return *peErr, nil
+	}
+
+	projectID := currentModel.ProjectId
+	instanceName := currentModel.InstanceName
+	connectionName := currentModel.ConnectionName
+	streamConnResp, apiResp, err := conn.StreamsApi.GetStreamConnection(context.Background(), *projectID, *instanceName, *connectionName).Execute()
+	if err != nil {
+		return handleError(apiResp, constants.READ, err)
+	}
+
+	resourceModel := GetStreamConnectionModel(streamConnResp, currentModel)
+
+	return handler.ProgressEvent{
+		OperationStatus: handler.Success,
+		ResourceModel:   resourceModel,
+	}, nil
+}
+
+func Update(req handler.Request, prevModel *Model, currentModel *Model) (handler.ProgressEvent, error) {
+	conn, peErr := setup(constants.UPDATE, req, currentModel)
+	if peErr != nil {
+		return *peErr, nil
+	}
+
+	ctx := context.Background()
+
+	projectID := currentModel.ProjectId
+	instanceName := currentModel.InstanceName
+	connectionName := currentModel.ConnectionName
+	streamConnectionReq := NewStreamConnectionReq(ctx, currentModel)
+	streamConnResp, apiResp, err := conn.StreamsApi.UpdateStreamConnection(ctx, *projectID, *instanceName, *connectionName, streamConnectionReq).Execute()
```

Comment:
just curious: you first extracted the params from `currentModel`, then you used `currentModel` to create the request and then you call the API passing both the request and the params. Is that how it works? or should you use the `prevModel`?

Reply:
yes, `projectId, instanceName & connectionName` are required to be passed as queryParams while remaining attributes are to be passed as the requestBody (`streamConnectionReq`). This is how the API is setup https://www.mongodb.com/docs/atlas/reference/api-resources-spec/v2/#tag/Streams/operation/updateStreamConnection